core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -237,605 +227,1223 @@ private SqlQueryScheduler(
+                // if there are no coordinator stages (e.g., simple select query) and the distributed stages are finished, do the query transitioning
+                // otherwise defer query transitioning to the coordinator stages
+                if (state == DistributedStagesSchedulerState.FINISHED) {
+                    queryStateMachine.transitionToFinishing();
+                }
+                else if (state == DistributedStagesSchedulerState.CANCELED) {
+                    // output stage was canceled
+                    queryStateMachine.transitionToCanceled();
+                }
+            }
+
+            if (state == DistributedStagesSchedulerState.FAILED) {
+                StageFailureInfo stageFailureInfo = distributedStagesScheduler.getFailureCause()
+                        .orElseGet(() -> new StageFailureInfo(toFailure(new VerifyException("distributedStagesScheduler failed but failure cause is not present")), Optional.empty()));
+                ErrorCode errorCode = stageFailureInfo.getFailureInfo().getErrorCode();
+                if (shouldRetry(errorCode)) {
+                    long delayInMillis = min(retryInitialDelay.toMillis() * ((long) pow(2, currentAttempt.get())), retryMaxDelay.toMillis());
+                    currentAttempt.incrementAndGet();
+                    scheduleRetryWithDelay(delayInMillis);
+                }
+                else {
+                    stageManager.getDistributedStagesInTopologicalOrder().forEach(stage -> {
+                        if (stageFailureInfo.getFailedStageId().isPresent() && stageFailureInfo.getFailedStageId().get().equals(stage.getStageId())) {
+                            stage.fail(stageFailureInfo.getFailureInfo().toException());
+                        }
+                        else {
+                            stage.abort();
+                        }
+                    });
+                    queryStateMachine.transitionToFailed(stageFailureInfo.getFailureInfo().toException());
+                }
+            }
+        });
+        return Optional.of(distributedStagesScheduler);
     }
 
-    private static Map<StageId, Set<StageId>> getStageLineage(QueryId queryId, SubPlan planTree)
+    private boolean shouldRetry(ErrorCode errorCode)
     {
-        ImmutableMap.Builder<StageId, Set<StageId>> result = ImmutableMap.builder();
-        for (SubPlan planNode : Traverser.forTree(SubPlan::getChildren).breadthFirst(planTree)) {
-            result.put(
-                    getStageId(queryId, planNode.getFragment().getId()),
-                    planNode.getChildren().stream()
-                            .map(stage -> getStageId(queryId, stage.getFragment().getId()))
-                            .collect(toImmutableSet()));
-        }
-        return result.build();
+        return retryPolicy == RetryPolicy.QUERY && currentAttempt.get() < maxRetryAttempts && (errorCode == null || errorCode.getType() != USER_ERROR);
```

Comment:
What about INSUFFICIENT_RESOURCES ?
It wouldn't help to retry the query if the number of nodes available in the cluster hasn't increased from last attempt.

Reply:
@linzebing also mentioned that. The criteria is not ideal. We should probably give it a little more thought. 

Currently it seems like the failure class that never makes sense to retry are `USER` failures (e.g.: division by zero). 

It is never 100% clear if it makes sense to retry an `INTERNAL` failure, as it may be caused by a bug? Yet there are probably some legit retriable failures in that class. 

Same for `EXTERNAL`. There are some failures that doesn't make sense to retry (e.g.: incorrect files layout for a bucketed table in Hive), yet there are many failures that are worth retrying.

Situation for `INSUFFICIENT_RESOURCES` is somehow similar. There are some failures that are not retriable (e.g.: local memory exceeded error). However there are some failures that are worth retrying (e.g.: `CLUSTER_OUT_OF_MEMORY`).

I think ideally we should extend error codes with a `retry` marker and mark all errors accordingly. However I would prefer having a separate PR for that. Let me add an item to https://github.com/trinodb/trino/issues/9101

Reply:
Local memory exceeded error is retryable in an autoscaling setup. But we need to know the number of nodes at start of previous attempt to make an informed decision about retrying it.
When running on spot nodes, the error codes which are useful to retry on in case of node loss are REMOTE_HOST_GONE, TOO_MANY_REQUESTS_FAILED and PAGE_TRANSPORT_TIMEOUT.
The problem with casting a wide net is that non-recoverable failures could take much longer to propagate back to user due to delays from retries. It could also result in unnecessary wastage of cluster resources if the query is resource hungry.
Maybe we can consider starting with a small targeted list of error codes to retry on and add to it as we discover more useful scenarios for retry.
I'm also wondering what's the behaviour now for query level limits like query.max-run-time, query.max-cpu-time, query.max-scan-physical-bytes etc. Does each attempt track and enforce these limits separately, or all the attempts put together need to stay under those limits ? If it's the latter then we could have the first attempt use up max-run-time and then subsequent attempts fail immediately after retry delays.

Reply:
> Local memory exceeded error is retryable in an autoscaling setup. But we need to know the number of nodes at start of previous attempt to make an informed decision about retrying it.
> The problem with casting a wide net is that non-recoverable failures could take much longer to propagate back to user due to delays from retries. It could also result in unnecessary wastage of cluster resources if the query is resource hungry.

That's a good point. 

> When running on spot nodes, the error codes which are useful to retry on in case of node loss are REMOTE_HOST_GONE, TOO_MANY_REQUESTS_FAILED and PAGE_TRANSPORT_TIMEOUT.

Ideally I wouldn't want to limit retries only to these errors as it limits the usefulness of the retry mechanism.

> Maybe we can consider starting with a small targeted list of error codes to retry on and add to it as we discover more useful scenarios for retry.

I'm leaning towards making `INSUFFICIENT_RESOURCES` type of errors as non retriable for now (until we revisit all error codes and mark retriable ones explicitly), as it feels like the downsides of retrying those errors outweighs the upsides. What do you think?

> I'm also wondering what's the behaviour now for query level limits like query.max-run-time, query.max-cpu-time, query.max-scan-physical-bytes etc. Does each attempt track and enforce these limits separately, or all the attempts put together need to stay under those limits ? If it's the latter then we could have the first attempt use up max-run-time and then subsequent attempts fail immediately after retry delays.

That's another very good question. Thanks for bringing it up. That I believe something we have to decide. I haven't had a chance to think more about how and where to account failures. This is one of the items on my list.

Reply:
> I'm leaning towards making INSUFFICIENT_RESOURCES type of errors as non retriable for now (until we revisit all error codes and mark retriable ones explicitly), as it feels like the downsides of retrying those errors outweighs the upsides. What do you think?

Agreed, though we can keep retry for CLUSTER_OUT_OF_MEMORY as that seems safe to retry and would probably be useful in some scenarios.

Reply:
Updated